Makefile

```diff
@@ -8,268 +28,428 @@ usage:
-	@printf "$$(tput bold)*************$$(tput sgr0)\n"
-	@for f in `./tools/list_archs.sh`; do echo "$$(tput bold)Test $$f"; cd arch/$$f; CI=true RUSTFLAGS="-D warnings" TOCK_KERNEL_VERSION=ci_test cargo test || exit 1; cd ../..; done
-
-.PHONY: ci-chips
-ci-chips:
-	@printf "$$(tput bold)*************$$(tput sgr0)\n"
-	@printf "$$(tput bold)* CI: Chips *$$(tput sgr0)\n"
-	@printf "$$(tput bold)*************$$(tput sgr0)\n"
-	@for f in `./tools/list_chips.sh`; do echo "$$(tput bold)Test $$f"; cd chips/$$f; CI=true RUSTFLAGS="-D warnings" TOCK_KERNEL_VERSION=ci_test cargo test || exit 1; cd ../..; done
-
-.PHONY: ci-kernel
-ci-kernel:
-	@printf "$$(tput bold)**************$$(tput sgr0)\n"
-	@printf "$$(tput bold)* CI: Kernel *$$(tput sgr0)\n"
-	@printf "$$(tput bold)**************$$(tput sgr0)\n"
-	@cd kernel && CI=true RUSTFLAGS="-D warnings" TOCK_KERNEL_VERSION=ci_test cargo test
 
-.PHONY: ci-syntax
-ci-syntax:
-	@printf "$$(tput bold)**************$$(tput sgr0)\n"
-	@printf "$$(tput bold)* CI: Syntax *$$(tput sgr0)\n"
-	@printf "$$(tput bold)**************$$(tput sgr0)\n"
+### ci-runner-github-build jobs:
+.PHONY: ci-job-syntax
+ci-job-syntax:
+	$(call banner,CI-Job: Syntax)
 	@CI=true $(MAKE) allcheck
 
-.PHONY: ci-compilation
-ci-compilation:
-	@printf "$$(tput bold)*******************$$(tput sgr0)\n"
-	@printf "$$(tput bold)* CI: Compilation *$$(tput sgr0)\n"
-	@printf "$$(tput bold)*******************$$(tput sgr0)\n"
+.PHONY: ci-job-compilation
+ci-job-compilation:
+	$(call banner,CI-Job: Compilation)
 	@CI=true $(MAKE) allboards
 
-.PHONY: ci-debug-support-targets
-ci-debug-support-targets:
+.PHONY: ci-job-debug-support-targets
+ci-job-debug-support-targets:
+	$(call banner, CI-Job: Debug Support Targets)
 	# These are rules that build additional debugging information, but are
 	# also quite time consuming. So we want to verify that the rules still
 	# work, but don't build them for every board.
 	#
 	# The choice of building for the nrf52dk was chosen by random die roll.
-	@printf "$$(tput bold)*****************************$$(tput sgr0)\n"
-	@printf "$$(tput bold)* CI: Debug Support Targets *$$(tput sgr0)\n"
-	@printf "$$(tput bold)*****************************$$(tput sgr0)\n"
 	@CI=true $(MAKE) -C boards/nordic/nrf52dk lst
 	@CI=true $(MAKE) -C boards/nordic/nrf52dk debug
 	@CI=true $(MAKE) -C boards/nordic/nrf52dk debug-lst
 
-.PHONY: ci-documentation
-ci-documentation:
-	@printf "$$(tput bold)*********************$$(tput sgr0)\n"
-	@printf "$$(tput bold)* CI: Documentation *$$(tput sgr0)\n"
-	@printf "$$(tput bold)*********************$$(tput sgr0)\n"
-	@CI=true tools/toc.sh
-
-
-## Actual Rules (Netlify)
+.PHONY: ci-job-collect-artifacts
+ci-job-collect-artifacts: ci-job-compilation
+	# Collect binary images for each board
+	#
+	# This is currently used only for code size detection changes, but in
+	# the future may also be used to support checks for deterministic builds.
+	@mkdir -p ./ci-artifacts
+	@rm -rf "./ci-artifacts/*"
```

Comment:
Wouldn't it be clearer to first remove the folder and then create it, instead of creating it and removing every file in it? The rm rule would be simpler: `rm -rf ci-artifacts`.

I also don't get why some paths start with `./` and some not (throughout this Makefile). Isn't this prefix useful only when running an executable?

Reply:
I've updated this to wipe and re-create the folder. I've also moved the folder into `tools/` to avoid cluttering the root if run locally.

As for `./` I assume that was just style of the people who wrote rules. I've updated such that its' only used for executables now.